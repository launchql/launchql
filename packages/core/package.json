{
  "name": "@launchql/core",
  "version": "2.11.0",
  "author": "Dan Lynch <pyramation@gmail.com>",
  "description": "LaunchQL Package and Migration Tools",
  "main": "index.js",
  "module": "esm/index.js",
  "types": "index.d.ts",
  "homepage": "https://github.com/launchql/launchql",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/launchql/launchql"
  },
  "bugs": {
    "url": "https://github.com/launchql/launchql/issues"
  },
  "scripts": {
    "copy": "npm run copy:pkg; npm run copy:sql",
    "copy:pkg": "copyfiles -f ../../LICENSE README.md package.json dist",
    "copy:sql": "copyfiles -f src/migrate/sql/* dist/migrate/sql && copyfiles -f src/migrate/sql/* dist/esm/migrate/sql && copyfiles -f src/init/sql/* dist/init/sql && copyfiles -f src/init/sql/* dist/esm/init/sql",
    "clean": "rimraf dist/**",
    "prepare": "npm run build",
    "build": "npm run clean; tsc; tsc -p tsconfig.esm.json; npm run copy",
    "build:dev": "npm run clean; tsc --declarationMap; tsc -p tsconfig.esm.json; npm run copy",
    "lint": "eslint . --fix",
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "keywords": [
    "database",
    "migration",
    "postgresql",
    "launchql",
    "schema",
    "sqitch"
  ],
  "devDependencies": {
    "@types/glob": "^8.1.0",
    "@types/rimraf": "^4.0.5"
  },
  "dependencies": {
    "@launchql/env": "^2.4.0",
    "@launchql/server-utils": "^2.4.0",
    "@launchql/templatizer": "^2.3.0",
    "@launchql/types": "^2.5.0",
    "@pgsql/types": "^17.6.1",
    "case": "^1.6.3",
    "chalk": "^4.1.0",
    "csv-to-pg": "^2.0.10",
    "glob": "^11.0.2",
    "parse-package-name": "^1.0.0",
    "pg-cache": "^1.3.0",
    "pg-env": "^1.1.0",
    "pgsql-deparser": "^17.11.1",
    "pgsql-parser": "^17.8.2",
    "rimraf": "^6.0.1"
  },
  "peerDependencies": {
    "@pgsql/types": "^17.6.1"
  }
}
