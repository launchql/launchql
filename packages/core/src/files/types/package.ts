export interface PackageAnalysisIssue {
  code: string;
  message: string;
  file?: string;
  fixable?: boolean;
}

export interface PackageAnalysisResult {
  ok: boolean;
  name: string;
  path: string;
  issues: PackageAnalysisIssue[];
}

export interface RenameOptions {
  dryRun?: boolean;
  syncPackageJsonName?: boolean;
}

export interface RenameResult {
  changed: string[];
  warnings: string[];
}
