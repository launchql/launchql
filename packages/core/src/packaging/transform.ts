import { deparse } from 'pgsql-deparser';
import { parse } from 'pgsql-parser';

type TransformProps = {
  [key: string]: ((value: any) => any) | { [key: string]: any };
};

/**
 * Recursively transforms the properties of an object based on a transformation map.
 * 
 * @param obj - The object to transform.
 * @param props - A map of properties and their transformation rules.
 * @returns A new object with transformed properties.
 */
export const transformProps = (obj: any, props: TransformProps): any => {
  let copy: any;

  // Handle primitive types, null, or undefined
  if (obj === null || typeof obj !== 'object') return obj;

  // Handle Date
  if (obj instanceof Date) {
    copy = new Date();
    copy.setTime(obj.getTime());
    return copy;
  }

  // Handle Array
  if (Array.isArray(obj)) {
    copy = [];
    for (let i = 0, len = obj.length; i < len; i++) {
      copy[i] = transformProps(obj[i], props);
    }
    return copy;
  }

  // Handle Object
  if (obj instanceof Object || typeof obj === 'object') {
    copy = {};
    for (const attr in obj) {
      if (Object.prototype.hasOwnProperty.call(obj, attr)) {
        if (props.hasOwnProperty(attr)) {
          const propRule = props[attr];
          if (typeof propRule === 'function') {
            // Apply function transformation
            copy[attr] = propRule(obj[attr]);
          } else if (typeof propRule === 'object' && propRule !== null) {
            // Apply value-based transformation
            if (propRule.hasOwnProperty(obj[attr])) {
              copy[attr] = propRule[obj[attr]];
            } else {
              copy[attr] = transformProps(obj[attr], props);
            }
          }
        } else {
          copy[attr] = transformProps(obj[attr], props);
        }
      }
    }
    return copy;
  }

  throw new Error("Unable to copy obj! Its type isn't supported.");
};

/**
 * Parses a SQL statement, transforms its properties, and regenerates the SQL.
 * 
 * @param statement - The SQL statement to transform.
 * @param props - A map of properties and their transformation rules.
 * @returns The transformed SQL statement.
 */
export const transform = async (statement: string, props: TransformProps): Promise<string> => {
  let tree = await parse(statement);
  tree = transformProps(tree, props);
  return await deparse(tree as any);
};
