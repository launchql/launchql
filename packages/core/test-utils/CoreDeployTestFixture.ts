import { getEnvOptions } from '@launchql/env';
import { join } from 'path';
import { teardownPgPools } from 'pg-cache';
import { getPgEnvOptions } from 'pg-env';

import { LaunchQLProject } from '../src/core/class/launchql';
import { MigrateTestFixture, TestDatabase } from './index';
import { TestFixture } from './TestFixture';

export class CoreDeployTestFixture extends TestFixture {
  private databases: TestDatabase[] = [];
  private migrateFixture?: MigrateTestFixture;

  constructor(...fixturePath: string[]) {
    super(...fixturePath);
    this.migrateFixture = new MigrateTestFixture();
  }

  async setupTestDatabase(): Promise<TestDatabase> {
    if (!this.migrateFixture) {
      throw new Error('MigrateTestFixture not initialized');
    }
    
    const db = await this.migrateFixture.setupTestDatabase();
    this.databases.push(db);
    return db;
  }

  async deployModule(projectName: string, database: string, fixturePath: string[], logOnly: boolean = false): Promise<void> {
    const basePath = this.tempFixtureDir;
    const originalCwd = process.cwd();
    
    try {
      process.chdir(basePath);
      
      const project = new LaunchQLProject(basePath);
      
      const opts = getEnvOptions({ 
        pg: getPgEnvOptions({ database }),
        deployment: {
          fast: false,
          usePlan: true,
          logOnly
        }
      });

      await project.deploy(opts, projectName, undefined, true);
    } finally {
      process.chdir(originalCwd);
    }
  }

  async revertModule(projectName: string, database: string, fixturePath: string[], toChange?: string): Promise<void> {
    const basePath = this.tempFixtureDir;
    const originalCwd = process.cwd();
    
    try {
      const projectPath = join(basePath, 'packages', projectName);
      process.chdir(projectPath);
      
      const project = new LaunchQLProject(projectPath);
      
      const opts = getEnvOptions({ 
        pg: getPgEnvOptions({ database })
      });

      await project.revert(opts, projectName, toChange, true);
    } finally {
      process.chdir(originalCwd);
    }
  }

  async verifyModule(projectName: string, database: string, fixturePath: string[], toChange?: string): Promise<void> {
    const basePath = this.tempFixtureDir;
    const originalCwd = process.cwd();
    
    try {
      process.chdir(basePath);
      
      const project = new LaunchQLProject(basePath);
      
      const opts = getEnvOptions({ 
        pg: getPgEnvOptions({ database })
      });

      await project.verify(opts, projectName, toChange, true);
    } finally {
      process.chdir(originalCwd);
    }
  }

  async cleanup(): Promise<void> {
    for (const db of this.databases) {
      await db.close();
    }
    this.databases = [];
    
    if (this.migrateFixture) {
      await this.migrateFixture.cleanup();
    }
    
    await teardownPgPools();
    
    // Small delay to ensure connections are fully closed
    await new Promise(resolve => setTimeout(resolve, 10));
    
    super.cleanup();
  }
}
