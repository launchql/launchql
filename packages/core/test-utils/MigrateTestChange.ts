export interface MigrateTestChange {
  name: string;
  dependencies?: string[];
  timestamp?: string;
  planner?: string;
  email?: string;
  comment?: string;
}
