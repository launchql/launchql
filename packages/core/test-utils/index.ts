export { TestPlan } from './TestPlan';
export { CoreDeployTestFixture } from './CoreDeployTestFixture';
export { TestFixture } from './TestFixture';

export { MigrateTestFixture } from './MigrateTestFixture';
export { TestDatabase } from './TestDatabase';
export { MigrateTestChange } from './MigrateTestChange';

export {
  FIXTURES_PATH,
  getFixturePath,
  cleanText,
  teardownAllPools,
  closeDatabasePools
} from './utils';
