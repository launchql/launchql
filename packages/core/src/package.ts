import { readFileSync, writeFileSync, mkdirSync, rmSync } from 'fs';
import { relative } from 'path';
import { parse } from 'pgsql-parser';
import { deparse } from 'pgsql-deparser';

import { getExtensionName } from './files';
import { resolve, resolveWithPlan } from './resolve';
import { transformProps } from './transform';
import { Logger } from '@launchql/logger';
import { RawStmt } from '@pgsql/types';

const log = new Logger('package');

const noop = (): any => undefined;

export const cleanTree = (tree: any): any => {
  return transformProps(tree, {
    stmt_len: noop,
    stmt_location: noop,
    location: noop,
  });
};

interface PackageModuleOptions {
  usePlan?: boolean;
  extension?: boolean;
  pretty?: boolean;
  functionDelimiter?: string;
}

interface WritePackageOptions extends PackageModuleOptions {
  version: string;
  packageDir: string;
}

const filterStatements = (stmts: RawStmt[], extension: boolean): RawStmt[] => {
  if (!extension) return stmts;
  return stmts.filter(node => {
    const stmt = node.stmt;
    return !stmt.hasOwnProperty('TransactionStmt') && 
           !stmt.hasOwnProperty('CreateExtensionStmt');
  });
};

export const packageModule = async (
  packageDir: string,
  { usePlan = true, extension = true, pretty = true, functionDelimiter = '$EOFCODE$' }: PackageModuleOptions = {}
): Promise<{ sql: string; diff?: boolean; tree1?: string; tree2?: string }> => {
  const resolveFn = usePlan ? resolveWithPlan : resolve;
  const sql = resolveFn(packageDir);

  if (!sql?.trim()) {
    log.warn(`⚠️ No SQL generated for module at ${packageDir}. Skipping.`);
    return { sql: '' };
  }

  const extname = getExtensionName(packageDir);

  try {
    const parsed = await parse(sql);
    parsed.stmts = filterStatements(parsed.stmts as any, extension);

    const topLine = extension
      ? `\\echo Use "CREATE EXTENSION ${extname}" to load this file. \\quit\n`
      : '';

    const finalSql = await deparse(parsed, {
      pretty,
      functionDelimiter
    });

    const tree1 = parsed.stmts;
    const tree2 = await parse(finalSql);

    const results: {
      sql: string;
      diff?: boolean;
      tree1?: string;
      tree2?: string;
    } = {
      sql: `${topLine}${finalSql}`,
    };

    const diff =
      JSON.stringify(cleanTree(tree1)) !== JSON.stringify(cleanTree(tree2));
    if (diff) {
      results.diff = true;
      results.tree1 = JSON.stringify(cleanTree(tree1), null, 2);
      results.tree2 = JSON.stringify(cleanTree(tree2), null, 2);
    }

    return results;
  } catch (e) {
    log.error(`❌ Failed to parse SQL for ${packageDir}`);
    console.error(e);
    throw e;
  }
};

export const writePackage = async ({
  version,
  extension = true,
  usePlan = true,
  packageDir,
}: WritePackageOptions): Promise<void> => {
  const pkgPath = `${packageDir}/package.json`;
  const pkg = require(pkgPath);
  const extname = await getExtensionName(packageDir);
  const makePath = `${packageDir}/Makefile`;
  const controlPath = `${packageDir}/${extname}.control`;
  const sqlFileName = `${extname}--${version}.sql`;

  const Makefile = readFileSync(makePath, 'utf-8');
  const control = readFileSync(controlPath, 'utf-8');

  const { sql, diff, tree1, tree2 } = await packageModule(packageDir, {
    extension,
    usePlan,
  });

  const outPath = extension ? `${packageDir}/sql` : `${packageDir}/out`;

  rmSync(outPath, { recursive: true, force: true });
  mkdirSync(outPath, { recursive: true });

  if (extension) {
    writeFileSync(
      controlPath,
      control.replace(
        /default_version = '[0-9\.]+'/,
        `default_version = '${version}'`
      )
    );

    pkg.version = version;
    writeFileSync(pkgPath, JSON.stringify(pkg, null, 2));

    const regex = new RegExp(`${extname}--[0-9.]+.sql`);
    writeFileSync(makePath, Makefile.replace(regex, sqlFileName));
  }

  if (diff) {
    log.warn(`⚠️ SQL diff exists! Review the ${relative(packageDir, outPath)}/ folder.`);
    // Uncomment if needed:
    // writeFileSync(`${outPath}/orig.${sqlFileName}.tree.json`, tree1);
    // writeFileSync(`${outPath}/parsed.${sqlFileName}.tree.json`, tree2);
  }

  const writePath = `${outPath}/${sqlFileName}`;
  writeFileSync(writePath, sql);
  log.success(`${relative(packageDir, writePath)} written`);
};
