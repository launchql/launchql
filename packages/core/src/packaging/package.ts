import { Logger } from '@launchql/logger';
import { RawStmt } from '@pgsql/types';
import { mkdirSync, readFileSync, rmSync, writeFileSync, existsSync } from 'fs';
import { relative } from 'path';
import { deparse } from 'pgsql-deparser';
import { parse } from 'pgsql-parser';

import { getExtensionName } from '../files';
import { resolve, resolveWithPlan } from '../resolution/resolve';
import { transformProps } from './transform';

const log = new Logger('package');

const noop = (): any => undefined;

export const cleanTree = (tree: any): any => {
  return transformProps(tree, {
    stmt_len: noop,
    stmt_location: noop,
    location: noop,
  });
};

interface PackageModuleOptions {
  usePlan?: boolean;
  extension?: boolean;
  pretty?: boolean;
  functionDelimiter?: string;
}

interface WritePackageOptions extends PackageModuleOptions {
  version: string;
  packageDir: string;
}

const filterStatements = (stmts: RawStmt[], extension: boolean): RawStmt[] => {
  if (!extension) return stmts;
  return stmts.filter(node => {
    const stmt = node.stmt;
    return !stmt.hasOwnProperty('TransactionStmt') && 
           !stmt.hasOwnProperty('CreateExtensionStmt');
  });
};

export const packageModule = async (
  packageDir: string,
  { usePlan = true, extension = true, pretty = true, functionDelimiter = '$EOFCODE$' }: PackageModuleOptions = {}
): Promise<{ sql: string; diff?: boolean; tree1?: string; tree2?: string }> => {
  const resolveFn = usePlan ? resolveWithPlan : resolve;
  const sql = resolveFn(packageDir);

  if (!sql?.trim()) {
    log.warn(`⚠️ No SQL generated for module at ${packageDir}. Skipping.`);
    return { sql: '' };
  }

  const extname = getExtensionName(packageDir);

  try {
    const parsed = await parse(sql);
    parsed.stmts = filterStatements(parsed.stmts as any, extension);

    const topLine = extension
      ? `\\echo Use "CREATE EXTENSION ${extname}" to load this file. \\quit\n`
      : '';

    const finalSql = await deparse(parsed, {
      pretty,
      functionDelimiter
    });

    const tree1 = parsed.stmts;
    const tree2 = await parse(finalSql);

    const results: {
      sql: string;
      diff?: boolean;
      tree1?: string;
      tree2?: string;
    } = {
      sql: `${topLine}${finalSql}`,
    };

    const diff =
      JSON.stringify(cleanTree(tree1)) !== JSON.stringify(cleanTree(tree2));
    if (diff) {
      results.diff = true;
      results.tree1 = JSON.stringify(cleanTree(tree1), null, 2);
      results.tree2 = JSON.stringify(cleanTree(tree2), null, 2);
    }

    return results;
  } catch (e) {
    log.error(`❌ Failed to parse SQL for ${packageDir}`);
    console.error(e);
    throw e;
  }
};

export const writePackage = async ({
  version,
  extension = true,
  usePlan = true,
  packageDir,
}: WritePackageOptions): Promise<void> => {
  const pkgPath = `${packageDir}/package.json`;
  const pkg = require(pkgPath);
  const extname = await getExtensionName(packageDir);
  const makePath = `${packageDir}/Makefile`;
  const controlPath = `${packageDir}/${extname}.control`;
  const sqlFileName = `${extname}--${version}.sql`;

  const hasMakefile = existsSync(makePath);
  const Makefile = hasMakefile ? readFileSync(makePath, 'utf-8') : undefined;
  const control = readFileSync(controlPath, 'utf-8');

  const { sql, diff, tree1, tree2 } = await packageModule(packageDir, {
    extension,
    usePlan,
  });

  const outPath = extension ? `${packageDir}/sql` : `${packageDir}/out`;

  rmSync(outPath, { recursive: true, force: true });
  mkdirSync(outPath, { recursive: true });

  if (extension) {
    writeFileSync(
      controlPath,
      control.replace(
        /default_version = '[0-9\.]+'/,
        `default_version = '${version}'`
      )
    );

    pkg.version = version;
    writeFileSync(pkgPath, JSON.stringify(pkg, null, 2));

    if (hasMakefile && Makefile) {
      const regex = new RegExp(`${extname}--[0-9.]+.sql`);
      writeFileSync(makePath, Makefile.replace(regex, sqlFileName));
    }
  }

  if (diff) {
    log.warn(`⚠️ SQL diff exists! Review the ${relative(packageDir, outPath)}/ folder.`);
    // Uncomment if needed:
    // writeFileSync(`${outPath}/orig.${sqlFileName}.tree.json`, tree1);
    // writeFileSync(`${outPath}/parsed.${sqlFileName}.tree.json`, tree2);
  }

  const writePath = `${outPath}/${sqlFileName}`;
  writeFileSync(writePath, sql);
  log.success(`${relative(packageDir, writePath)} written`);
};
