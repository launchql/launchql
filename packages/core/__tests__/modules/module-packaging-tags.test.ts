import { packageModule } from '../../src/packaging/package';
import {TestFixture } from '../../test-utils';

let fixtures: TestFixture[];

beforeAll(() => {
  fixtures = [
    new TestFixture('sqitch', 'simple-w-tags', 'packages', 'my-first'),
    new TestFixture('sqitch', 'simple-w-tags', 'packages', 'my-second'),
    new TestFixture('sqitch', 'simple-w-tags', 'packages', 'my-third'),
  ];
});

afterAll(() => {
  fixtures.forEach(fixture => fixture.cleanup());
});

it('my-first extension', async () => {
  const { sql } = await packageModule(fixtures[0].tempFixtureDir);
  expect(sql).toMatchSnapshot();
});

it('my-first extension via plan', async () => {
  const { sql } = await packageModule(fixtures[0].tempFixtureDir, { usePlan: true });
  expect(sql).toMatchSnapshot();
});

it('my-second extension', async () => {
  const { sql } = await packageModule(fixtures[1].tempFixtureDir);
  expect(sql).toMatchSnapshot();
});

it('my-second extension via plan', async () => {
  const { sql } = await packageModule(fixtures[1].tempFixtureDir, { usePlan: true });
  expect(sql).toMatchSnapshot();
});

it('my-third extension', async () => {
  const { sql } = await packageModule(fixtures[2].tempFixtureDir);
  expect(sql).toMatchSnapshot();
});

it('my-third extension via plan', async () => {
  const { sql } = await packageModule(fixtures[2].tempFixtureDir, { usePlan: true });
  expect(sql).toMatchSnapshot();
});
