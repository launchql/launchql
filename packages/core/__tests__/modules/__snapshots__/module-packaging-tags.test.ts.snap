// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my-first extension 1`] = `
"\\echo Use "CREATE EXTENSION my-first" to load this file. \\quit
CREATE SCHEMA IF NOT EXISTS myfirstapp;

CREATE TABLE myfirstapp.users (
  id serial PRIMARY KEY,
  username text NOT NULL UNIQUE,
  email text NOT NULL UNIQUE,
  created_at pg_catalog.timestamp DEFAULT now()
);

CREATE TABLE myfirstapp.products (
  id serial PRIMARY KEY,
  name text NOT NULL,
  description text,
  price numeric(10, 2) NOT NULL CHECK (price >= 0),
  in_stock boolean DEFAULT true,
  created_at timestamp with time zone DEFAULT now(),
  updated_at timestamp with time zone DEFAULT now()
);"
`;

exports[`my-first extension via plan 1`] = `
"\\echo Use "CREATE EXTENSION my-first" to load this file. \\quit
CREATE SCHEMA IF NOT EXISTS myfirstapp;

CREATE TABLE myfirstapp.users (
  id serial PRIMARY KEY,
  username text NOT NULL UNIQUE,
  email text NOT NULL UNIQUE,
  created_at pg_catalog.timestamp DEFAULT now()
);

CREATE TABLE myfirstapp.products (
  id serial PRIMARY KEY,
  name text NOT NULL,
  description text,
  price numeric(10, 2) NOT NULL CHECK (price >= 0),
  in_stock boolean DEFAULT true,
  created_at timestamp with time zone DEFAULT now(),
  updated_at timestamp with time zone DEFAULT now()
);"
`;

exports[`my-second extension 1`] = `
"\\echo Use "CREATE EXTENSION my-second" to load this file. \\quit
CREATE SCHEMA mysecondapp;

CREATE TABLE mysecondapp.users (
  id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
  email text UNIQUE NOT NULL,
  created_at timestamptz DEFAULT now(),
  last_active_at timestamptz
);

CREATE TABLE mysecondapp.user_interactions (
  id bigserial PRIMARY KEY,
  user_id uuid REFERENCES mysecondapp.users (id)
    ON DELETE CASCADE,
  interaction_type text NOT NULL CHECK (interaction_type IN ('click', 'hover', 'scroll', 'input')),
  target text NOT NULL,
  metadata jsonb,
  occurred_at timestamptz DEFAULT now()
);

CREATE TABLE mysecondapp.consent_agreements (
  id bigserial PRIMARY KEY,
  user_id uuid REFERENCES mysecondapp.users (id)
    ON DELETE CASCADE,
  consent_type text NOT NULL,
  granted_at timestamptz DEFAULT now(),
  revoked_at timestamptz,
  version text NOT NULL
);"
`;

exports[`my-second extension via plan 1`] = `
"\\echo Use "CREATE EXTENSION my-second" to load this file. \\quit
CREATE SCHEMA mysecondapp;

CREATE TABLE mysecondapp.users (
  id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
  email text UNIQUE NOT NULL,
  created_at timestamptz DEFAULT now(),
  last_active_at timestamptz
);

CREATE TABLE mysecondapp.user_interactions (
  id bigserial PRIMARY KEY,
  user_id uuid REFERENCES mysecondapp.users (id)
    ON DELETE CASCADE,
  interaction_type text NOT NULL CHECK (interaction_type IN ('click', 'hover', 'scroll', 'input')),
  target text NOT NULL,
  metadata jsonb,
  occurred_at timestamptz DEFAULT now()
);

CREATE TABLE mysecondapp.consent_agreements (
  id bigserial PRIMARY KEY,
  user_id uuid REFERENCES mysecondapp.users (id)
    ON DELETE CASCADE,
  consent_type text NOT NULL,
  granted_at timestamptz DEFAULT now(),
  revoked_at timestamptz,
  version text NOT NULL
);"
`;

exports[`my-third extension 1`] = `
"\\echo Use "CREATE EXTENSION my-third" to load this file. \\quit
CREATE SCHEMA mythirdapp;

CREATE TABLE mythirdapp.customers (
  id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
  email text UNIQUE NOT NULL,
  created_at timestamptz DEFAULT now(),
  last_active_at timestamptz
);"
`;

exports[`my-third extension via plan 1`] = `
"\\echo Use "CREATE EXTENSION my-third" to load this file. \\quit
CREATE SCHEMA mythirdapp;

CREATE TABLE mythirdapp.customers (
  id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
  email text UNIQUE NOT NULL,
  created_at timestamptz DEFAULT now(),
  last_active_at timestamptz
);"
`;
